### ISSUE TITLE
Assert.That with ItemsConstraintExpression calls IFormattable implementation

### ISSUE THREAD

**Issue by TobiasSekan** [nunit/nunit#3350]
Hello,

# Problem
when I use a list with objects that are implement the `System.IFormattable` interface
```csharp
using System;

public sealed class MyClass : IFormattable
{

    // ...

    public string ToString(string format, IFormatProvider formatProvider)
    {
        throw new FormatException();
    }
}
```

## Actual behaviour
the `ItemsConstraintExpression` fail with `FormatException`
```csharp

var myList = new List<MyClass>();

myList.Add(new MyClass());
myList.Add(new MyClass());

Assert.That(myList, Has.Exactly(2).Items);
```

## Expected behaviour
The constraint should only count the items and the test should pass

## Info
* Visual Studio Community 2019 Preview - Version 16.3.0 Preview 2.0
* Frameworks: .Net Standard 1.3 (program), .Net Core 2.2 (Unit-Test)
* NUnit 3.12.0 (NuGet package)
* NUnit3TestAdapter 3.14.0 (NuGet package)

**Comment by mikkelbu** [comment]
Hi @TobiasSekan 
Thanks for a very good bug description. Can you either upload a minimal example to github or just attach it to this issue and also provide a stacktrace of the failure?

The reason for this is that I cannot recreate the problem. When I try to mimick your setup then the test passes. 
![image](https://user-images.githubusercontent.com/1007631/63374068-5d9cbe00-c389-11e9-82fb-02953482414c.png)
How are the two projects connected? In my setup I have a `ProjectReference` from the test project to the prod project.

**Comment by TobiasSekan** [comment]
Ups my fault, you must expect more items as the list have

```csharp
using NUnit.Framework;
using System;
using System.Collections.ObjectModel;

namespace MyTestNamespace
{

    [TestFixture]
    public sealed class MainTestClass
    {
        public sealed class MyClass : IFormattable
        {
            public string ToString(string format, IFormatProvider formatProvider)
            {
                throw new FormatException();
            }
        }

        [Test]
        public void Test()
        {
            // only two items
            var myList = new Collection<MyClass>
            {
                new MyClass(),
                new MyClass()
            };

            // but we expect three items
            Assert.That(myList, Has.Exactly(3).Items);
        }
    }
}
```

```
System.FormatException : One of the identified items was in an invalid format.
   at MyTestNamespace.MainTestClass.MyClass.ToString(String format, IFormatProvider formatProvider)

   at System.Text.StringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.FormatHelper(IFormatProvider provider, String format, ParamsArray args)

   at NUnit.Framework.Constraints.MsgUtils.<>c.<.cctor>b__14_0(Object val)
   at NUnit.Framework.Constraints.MsgUtils.<>c__DisplayClass14_1.<.cctor>b__15(Object val)
   at NUnit.Framework.Constraints.MsgUtils.<>c__DisplayClass14_2.<.cctor>b__16(Object val)
   at NUnit.Framework.Constraints.MsgUtils.FormatCollection(IEnumerable collection, Int64 start, Int32 max)

   at NUnit.Framework.Constraints.ExactCountConstraintResult.WriteActualValueTo(MessageWriter writer)

   at NUnit.Framework.Internal.TextMessageWriter.WriteActualLine(ConstraintResult result)
   at NUnit.Framework.Internal.TextMessageWriter.DisplayDifferences(ConstraintResult result)

   at NUnit.Framework.Assert.ReportFailure(ConstraintResult result, String message, Object[] args)
   at NUnit.Framework.Assert.That[TActual](TActual actual, IResolveConstraint expression)

   at MyTestNamespace.MainTestClass.Test()
```

**Comment by mikkelbu** [comment]
The problem is that we always write the string representation of the elements (at least for the first 10 elements if I remember correctly) when the assert fails.

So if we have something similar to above, but where `MyClass` accepts an integer and then just return the integer as a string in `ToString(...)` then the following test 
```csharp
    public void ATest()
    {
        var myList = new List<MyClass>();

        myList.Add(new MyClass(1));
        myList.Add(new MyClass(2));
        myList.Add(new MyClass(44));

        Assert.That(myList, Has.Exactly(4).Items);
    }
```
would fail with 
```
      Expected: exactly 4 items
      But was:  3 items < <1>, <2>, <44> >
```
One could argue that it is unnecessary to write the individual elements, when we are only asserting on the number of items, but this constraint can be part of a more complex constraint. So if we have something that checks that some number of elements satisfies something
```csharp
public void ATest()
    {
        var myList = new List<MyClass>();

        myList.Add(new MyClass(1));
        myList.Add(new MyClass(2));
        var item = new MyClass(44);
        myList.Add(item);

        Assert.That(myList, Has.Exactly(2).Items.EqualTo(item));
    }
```
Then it can fail with
```
      Expected: exactly 2 items equal to <44>
      But was:  1 item < <1>, <2>, <44> >
```

For this reason I think we should close this issue as "working as intended", as I think adding special handling here would make the code more complex without much gain. @nunit/framework-and-engine-team what do you think?

Ps. If you just want to test on the length of the list, then you can use `Assert.That(myList, Has.Count.EqualTo(2));`, as this does not write the individual elements. The error message has this pattern
```
      Expected: property Count equal to 2
      But was:  3
```

**Comment by TobiasSekan** [comment]
> Ps. If you just want to test on the length of the list, then you can use Assert.That(myList, Has.Count.EqualTo(2));, as this does not write the individual elements. The error message has this pattern

I didn't know that yet.

The documentation under
https://github.com/nunit/docs/wiki/ExactCountConstraint
don't inform about this

**Comment by mikkelbu** [comment]
> I didn't know that yet.

I didn't either before this issue, but I knew that there are a lot of different constraints that can express - more or less -the same property, so I tried a couple of approaches.

And you are right about the documentation. It can always be improved, but people only have limited (spare) time, so this is often this is prioritized below fixing bugs and adding new functionality. I've added a "See also..." section to https://github.com/nunit/docs/wiki/ExactCountConstraint.
